### ISSUE TITLE
Handle app install

### ISSUE THREAD

**Issue by tumido** [open-services-group/peribolos-as-a-service#6]
https://github.com/open-services-group/peribolos-as-a-service/blob/fab17cbeb13e9ec6d95ffed4b7952089558583cd/architecture.md#user-subscribe

Depends on: https://github.com/open-services-group/peribolos-as-a-service/issues/4

1. Receives "install" webhook with "create" action
2. Fetches a token for the installation
3. Stores all data about the installation + a timestamp into a kubernetes secret

**Comment by KPostOffice** [comment]
I was thinking about this and I don't think that we need to store access tokens. Getting a new access token for each `peribolos` run shouldn't contribute significantly to the API limit. If we decide we do need to store it, I don't think we should use Kubernetes secrets, as creating a new secret requires additional permissions which I don't think we should add to the pod.

**Comment by tumido** [comment]
Ok, I think we need to break this down a bit.

> I was thinking about this and I don't think that we need to store access tokens. Getting a new access token for each peribolos run shouldn't contribute significantly to the API limit.

This really depends on when peribolos job is being triggered. Do we trigger on merge only? How often do we expect a merge to happen in a single org/installation?

Do we also want to support linting and dry runs on a PR? In that case each push to a pull request would trigger a new pipeline run. In case when we're not caching the tokens, we need to request a new token for every push to PR. This way a single user can get whole our service rate limited - a simple repeated git push in a single org can DOS the whole service for everybody... I don't think we want that to happen.

> If we decide we do need to store it, I don't think we should use Kubernetes secrets, as creating a new secret requires additional permissions which I don't think we should add to the pod.

Peribolos binary or tekton pipeline is not involved in generating tokens in either case. This is handled by the controller pod, which requires to access secrets anyways (because of app id, client secret, certificate...).

> additional permissions which I don't think we should add to the pod.

Any reason why a controller pod should not be given permissions to create a secret per app installataion? What's the concern here?

**Comment by KPostOffice** [comment]
> This really depends on when peribolos job is being triggered. Do we trigger on merge only? How often do we expect a merge to happen in a single org/installation?

Let me clarify here. When I say it won't contribute significantly I mean that, when compared to the number of API requests made by Peribolos will be much greater than the one used to get a new token for each Peribolos run.

> a simple repeated git push in a single org can DOS the whole service for everybody... I don't think we want that to happen.

Again, the limiting factor here is the Pipeline runs, not token generation.

> Any reason why a controller pod should not be given permissions to create a secret per app installataion? What's the concern here?

I'm just using the "Principle of Least Privilege". We shouldn't give the pod permissions it doesn't need to complete a task. If we do need to store the tokens, I feel it would be better to store them on a mounted PVC. It probably doesn't even need to be a PVC as losing tokens isn't an issue.

**Comment by tumido** [comment]
> I'm just using the "Principle of Least Privilege". We shouldn't give the pod permissions it doesn't need to complete a task. If we do need to store the tokens, I feel it would be better to store them on a mounted PVC. It probably doesn't even need to be a PVC as losing tokens isn't an issue.


Hm, I don't see how we can do this efficiently without creating secrets though. What we need to weight is:

#### 1) Who generates the token. Is it the controller? Or is it the peribolos job/task?

If we say we want it to be generated by the peribolos job itself, we need to pass an installation specific task credentials to whole github app (certificate, client secret etc.) which is shared across all installation, therefore potentially compromising all installations.

If we decide that we want it to be generated by the app controller, we need a way to pass the token to the peribolos task.

I think generating the token in the app controller and passing only an installation specific token to peribolos task is the more secure approach here.

#### 2) How do we want the token to be passed to the peribolos task from the controller?

I think we have two options here - a secret and PVC (suggested in your previous comment). Let's break it down again:

A secret is a kubernetes native way to pass sensitive data to pods. It doesn't inline the values like configmap does etc. Data is available either as environment variables or mounted filesystem. That makes the data available in runtime only and again limiting the possibility of leaking it through kubernetes API or persisting it in filesystem.

A PVC is a filesystem unit stored in a remote filesystem without any access control or auditing. A kubernetes user has no control over who else has access to the PV outside of the kubernetes cluster. Data is persisted. This is not ideal for sensitive data. Furthermore, the data exchange is limited here. Either we generate a new PVC per each installation and mount them all to the controller (would require reload on every install/uninstall) or we store all tokens in a single PVC making them available to all instances of the peribolos task, effectively sharing the credentials between all orgs. In addition to that, we also limit our scaling factor, since only few kubernetes PV backends support attaching PVs to different nodes at the same time (ReadWriteMany), hence limiting whole peribolos service to a single node within kubernetes cluster. This solution presents very big risks of leaking credentials and limits scaling, I don't think this is a good solution.

I really don't think being able to create secrets by a github app controller which has admin access to many github organizations creates any security risk or goes against the principle of least privilege. I think this is precisely the opposite. Just for the sake of not managing secrets in a namespace we would create much greater risks of leaking the credentials because we would either need to share the installation specific credentials in plaintext (inlining as env variable) or between installations + other unknowns entities (PVC) or even share the app credentials which which give access to all installations...

> Let me clarify here. When I say it won't contribute significantly I mean that, when compared to the number of API requests made by Peribolos will be much greater than the one used to get a new token for each Peribolos run.

That is true and it make sense. However due to the above I think we still should store and pass the token via a secret. And in that case we need to avoid race condition  when creating/updating the secret across multiple runs within each installation. I think in such case we're best at controlling when the credentials are generated.

**Comment by KPostOffice** [comment]
> I think generating the token in the app controller and passing only an installation specific token to peribolos task is the more secure approach here.

I completely agree with this. The main thing I'm not understanding is, what's keeping us from passing the token directly as an argument to the tekton task instead of inlining it as a secret?

> A PVC is a filesystem unit stored in a remote filesystem without any access control or auditing. A kubernetes user has no control over who...

When I suggested storing the tokens on disk it was under the assumption that they were only accessed directly by the controller and passed as args to the tekton task. I agree that requiring the peribolos task to mount a PVC is a bad idea.

**Comment by tumido** [comment]
I see. Well.. let's examine what happens in either case:

### Token as a parameter

Let's say our `TaskRun` resource look like this:

```yaml
apiVersion: tekton.dev/v1beta1
kind: TaskRun
metadata:
  generateName: peribolos-
spec:
  params:
    - name: token
      value: "THIS IS A SECRET"
  taskSpec:
    params:
    - name: token
      type: string
    steps:
    - name: default
      image: ...
      script: |
        echo $(params.token)
``` 

That means we've inlined the secret value into the `TaskRun` resource and everybody who has access to the namespace can read it. Since `TaskRun` resource doesn't have any special protections, it's really accessible to everybody with minimal permissions. Including different cluster scope scrapers and auditing tools. Basically everybody with etcd/api access. That's not very secure...

### Using secrets

Let's say our `TaskRun` definition looks like this now:

```yaml
apiVersion: tekton.dev/v1beta1
kind: TaskRun
metadata:
  generateName: peribolos-
spec:
  params:
    - name: installationId
      value: 123
  taskSpec:
    params:
    - name: installationId
      type: string
    steps:
    - name: default
      image: ...
      script: |
        echo $TOKEN
      env:
        - name: TOKEN
          valueFrom:
            secretKeyRef:
              name: $(params.installationId)
              key: token
      
``` 

Now everybody with access to the `TaskRun` resource/api can't get the token value from reading the manifest. They can't read it from any of the generated manifests either, like the `Pod` manifest. That is good.

However, you might argue that the same can apply to the `Secret` resource - anybody with access can read the secret. Yes, that is true, with a big exception. Reading a secret requires different level of access than reading regular resources. `Secret` resources in kubernetes are meant to store all sensitive data for a reason. `Secret` is a special kind of resource that is treated differently from all other resource types. For example simple view access in a namespace is not enough to view a `Secret`. Even `cluster-reader` role on cluster scope doesn't have access to secrets in individual namespaces. User needs to have at least edit access to namespaces to view secrets. Additionally, in Kubernetes all resources in etcd are stored in plaintext. By default all of them are in plaintext (bad practice if you ask me), however encryption can be enabled for some resources, see [here for OpenShift](https://docs.openshift.com/container-platform/4.7/security/encrypting-etcd.html), in vanilla Kubernetes, this is [advertised for secrets only, but can be extended](https://kubernetes.io/docs/tasks/administer-cluster/encrypt-data/).

In the end everybody with access to the individual pod or pod creation is able to exploit different Kubernetes ways to find and read secrets, that is true. On the other hand, `Secret`s try to limit sensitive data exposure on the api server level. It's a best practice in Kubernetes. 

Some good read on Kubernetes secrets is available in the first paragraphs of upstream docs:
https://kubernetes.io/docs/concepts/configuration/secret/

Also the "Secrets should be secrets" in this Oracle article:
https://blogs.oracle.com/developers/post/5-best-practices-for-kubernetes-security

The best resource I have on this is this RH blog article:
https://enterprisersproject.com/article/2019/8/kubernetes-secrets-explained-plain-english?extIdCarryOver=true&sc_cid=701f2000001Css5AAC&page=0%2C1

**Comment by KPostOffice** [comment]
Got it, thanks for the thorough reply @tumido. I had never seen the pattern of making secrets as needed and it  just was sitting wrong with me for some reason. Giving  the controller access to secret creation in order to make the tokens opaque to as many roles as possible makes sense.

**Comment by tumido** [comment]
Per installation secret should look like this:

```yaml
...
metadata:
  name: peribolos-<installation_id>
  annotations:
    expiresAt: <current_timestamp + 40min>
stringData:
  token: <token>
  orgName: <org_name>
```

**Comment by tumido** [comment]
@KPostOffice it just occured to me, you may need some help/guidance on the Kubernetes interface from the controller. Let me share this hint with you (if you have a different approach/idea in mind, don't hesitate and use it instead). In Meteor we use https://www.npmjs.com/package/@kubernetes/client-node npm package. Here's how we interfaces with it:

https://github.com/AICoE/meteor/blob/main/src/k8s.ts 

you'd need to use something like:
https://github.com/AICoE/meteor/blob/f45998ea99f3d9197d1c359313130b28ab1ce3c0/src/k8s.ts#L4-L17

```js
import fs from 'fs';
const k8s = require('@kubernetes/client-node');

const kc = new k8s.KubeConfig();[Tomas Coufal, 8 months ago: • feat: make shower next.js only](https://sourcegraph.com/github.com/AICoE/meteor/-/commit/7d6a13fcd7c71f03092dcd952f951807f0a101fe)
kc.loadFromDefault();

const k8sCustomApi = kc.makeApiClient(k8s.CustomObjectsApi);
const k8sCoreApi = kc.makeApiClient(k8s.CoreV1Api);
const k8sContext = kc.getCurrentContext();

const namespace = (() => {
  if (k8sContext === 'inClusterContext') {
    return fs.readFileSync('/var/run/secrets/kubernetes.io/serviceaccount/namespace', 'utf8');
  } else {
    return k8sContext.split('/')[0];
  }
})();


k8sCoreApi.createNamespacedSecret(
  namespace,
  {
    apiVersion: "v1",
    kind: "Secret",
    metadata: {
      name: ...
    },
    stringData: {
      token: ...
    },
  }
)
```

**Comment by harshad16** [comment]
> Per installation secret should look like this:
> 
> ```yaml
> ...
> metadata:
>   name: peribolos-<installation_id>
>   annotations:
>     expiresAt: <current_timestamp + 40min>
> stringData:
>   token: <token>
>   orgName: <org_name>
> ```

let's also add another annotations `component: peribolos-as-service`
for the easy filter of secrets while checking each secret in a namespace, for updated token.

**Comment by tumido** [comment]
Updated above. It can't be an annotation, but it needs to be a `label`. Kubernetes can't filter on `annotations`, only `labels`. Let's also use [some convention here](https://kubernetes.io/docs/concepts/overview/working-with-objects/common-labels/) and use `app.kubernetes.io/created-by: peribolos` here.

https://kubernetes.io/docs/concepts/overview/working-with-objects/annotations/#attaching-metadata-to-objects
https://blog.getambassador.io/kubernetes-labels-vs-annotations-95fc47196b6d

**Comment by tumido** [comment]
Resolved via #19
